### PR TITLE
CUMULUS-4147: Added function findDuplicateGranules return duplicate granules

### PR DESCRIPTION
**Pull request by jennyhliu**
**Summary:** Summary of changes

Addresses [CUMULUS-4147: Design and implement Granule Duplicate Detection function for "Active" Granules](https://bugs.earthdata.nasa.gov/browse/CUMULUS-4147)

## Changes

* Detailed list or prose of changes
* ...

## PR Checklist

- [ ] Update CHANGELOG
- [ ] Unit tests
- [ ] Ad-hoc testing - Deploy changes and test manually
- [ ] Integration tests

### CODE REVIEW

File: packages/db/tests/lib/test-duplicate-detection.js [hunk 1]
```diff
@@ -0,0 +1,221 @@
+const test = require('ava');
```
Comment (Jkovarik) [inline packages/db/tests/lib/test-duplicate-detection.js:1]: Rename to : test-duplicate-detection (after review comments are addressed so changeset doesn't become impossible to track)

File: packages/db/src/lib/duplicate-detection.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    (g) => g.collection_cumulus_id === collectionCumulusId
+  );
+  const differentCollectionMatches = duplicates.filter(
+    (g) => g.collection_cumulus_id !== collectionCumulusId
+  );
+
+  return {
+    sameCollectionMatches,
+    differentCollectionMatches,
+    customCriteriaMatches: [], // Placeholder
```
Comment (Jkovarik) [inline packages/db/src/lib/duplicate-detection.ts]: Nit: can we have a bit more context in this comment, or a link to the discussion in this PR?

File: packages/db/src/lib/duplicate-detection.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  group_state: string | null;
+};
+
+type DuplicateGranulesResult = {
+  sameCollectionMatches: GranuleRecord[];
+  differentCollectionMatches: GranuleRecord[];
+  customCriteriaMatches: GranuleRecord[];
+};
+
+export const getNextGranuleGroupId = async (knex: Knex) : Promise<number> => {
```
Comment (Jkovarik) [inline packages/db/src/lib/duplicate-detection.ts:39]: Is this the best location for this method?   Should it go in a lib/granule-groups.ts, or the model directly?

File: packages/db/src/lib/duplicate-detection.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
+import { Knex } from 'knex';
+import { deconstructCollectionId } from '@cumulus/message/Collections';
+import { CollectionPgModel } from '../models/collection';
+
+type GranuleInput = {
+  collectionId: string;
```
Comment (Jkovarik) [inline packages/db/src/lib/duplicate-detection.ts:6]: Should we do something like 

```
declare const NonEmptyBrand: unique symbol;
export type NonEmptyString = string & { readonly [NonEmptyBrand]: true };
```

for the required strings and require clients to validate the appropriate value, and rely on our jsdoc/ts-check for JS clients, *or* do a boundary condition check on incoming to make sure the input object has appropriate values to ensure we have useful error output given the db package is an api boundary? 

My gut says doing a quick runtime check is probably preferable here.

File: packages/db/src/lib/duplicate-detection.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      deconstructCollectionId(collectionId)
+    );
+    collectionCumulusId = pgCollection.cumulus_id;
+  }
+
+  // Query for active granules with matching producerGranuleId
+  const duplicates = await knex('granules')
+    .leftJoin('granule_groups', 'granules.cumulus_id', 'granule_groups.granule_cumulus_id')
+    .where('granules.producer_granule_id', producerGranuleId)
+    .modify((qb) => {
```
Comment (Nnaga1) [inline packages/db/src/lib/duplicate-detection.ts:85]: what does this do?, why does granule_id matter outside of the join?:
```
    .modify((qb) => {
      if (granuleId) {
        qb.whereNot('granules.granule_id', granuleId);
      }
    })
```

File: packages/db/src/lib/duplicate-detection.ts (hunk 2 above)
Comment (Nnaga1) [inline packages/db/src/lib/duplicate-detection.ts]: 👍

File: packages/db/src/migrations/20250915152638_create_granule_groups_table_and_index.ts [hunk 6]
```diff
@@ -11,7 +16,8 @@ export const up = async (knex: Knex): Promise<void> => {
     table
       .integer('group_id')
       .comment('Granule duplicate group id')
-      .notNullable();
+      .notNullable()
+      .defaultTo(knex.raw('nextval(\'granule_group_id_seq\')'));
```
Comment (Nnaga1) [inline packages/db/src/migrations/20250915152638_create_granule_groups_table_and_index.ts:20]: this is cool! didn't know sql could even have a sequential number like this

File: packages/db/src/lib/duplicate-detection.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+import { CollectionPgModel } from '../models/collection';
+
+type GranuleInput = {
+  collectionId: string;
+  producerGranuleId: string;
+  collectionCumulusId?: number;
+  granuleId?: string;
+};
+
+type GranuleRecord = {
```
Comment (Jkovarik) [inline packages/db/src/lib/duplicate-detection.ts]: Can we call this something other than a granule record?  It's not really a *granule* record, so much as a ....granuleGroupRecord?  

 Should this should be exported, in case we need to import it for use in our .js?

File: packages/db/tests/lib/test-dupliate-detection.js [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+const insertGranulesWithProducerId = async ({
+  knex,
+  granulePgModel,
+  count,
+  collectionCumulusId,
+  producerGranuleId,
+  granuleIdPrefix = 'g',
+}) => await granulePgModel.insert(
+  knex,
+  Array.from({ length: count }).map((_, i) =>
```
Comment (Nnaga1) [inline packages/db/tests/lib/test-dupliate-detection.js]: could you use `range` here instead? like: https://github.com/nasa/cumulus/blob/master/packages/db/tests/search/test-GranuleSearch.js#L131

File: packages/db/src/lib/duplicate-detection.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+export const findDuplicateGranules = async (
+  incoming: GranuleInput,
+  knex: Knex,
+  collectionPgModel = new CollectionPgModel()
+): Promise<DuplicateGranulesResult> => {
+  const {
+    collectionId,
+    producerGranuleId,
+    granuleId,
+  } = incoming;
```
Comment (Jkovarik) [inline packages/db/src/lib/duplicate-detection.ts]: Nit: prefer input over incoming

File: packages/db/src/lib/duplicate-detection.ts (hunk 2 above)
Comment (Jkovarik) [inline packages/db/src/lib/duplicate-detection.ts]: Please elaborate a bit re: the comment, or a direct link to the PR (for ease of tracking down what it's placeholding in the future), or something that avoids a history lookup on the commit

File: packages/db/tests/lib/test-duplicate-detection.js [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+
+  const producerGranuleId = 'same-producer-id-same-collection';
+
+  await insertGranulesWithProducerId({
+    knex,
+    granulePgModel,
+    count: 3,
+    collectionCumulusId,
+    producerGranuleId,
+    granuleIdPrefix: 'same-coll',
```
Comment (Nnaga1) [inline packages/db/tests/lib/test-duplicate-detection.js:102]: is this `granuleIdPrefix: 'same-coll',` used somewhere?

File: packages/db/src/lib/duplicate-detection.ts [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+  customCriteriaMatches: GranuleRecord[];
+};
+
+export const getNextGranuleGroupId = async (knex: Knex) : Promise<number> => {
+  const result = await knex.raw("SELECT nextval('granule_group_id_seq')::int");
+  return result.rows[0].nextval;
+};
+
+/**
+ * Detects "active" duplicate granules based on producerGranuleId and collection.
```
Comment (Jkovarik) [inline packages/db/src/lib/duplicate-detection.ts]: Docstring should to explain what the intended behaviors are for the values in GranuleInput, particularly the optional ones.

File: packages/db/tests/lib/test-duplicate-detection.js [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+  await setGranuleGroupStates(knex, activeGranules, 'A');
+
+  const result = await findDuplicateGranules({
+    collectionId,
+    producerGranuleId,
+    collectionCumulusId,
+  }, knex);
+
+  t.is(result.sameCollectionMatches.length, 2);
+  t.true(result.sameCollectionMatches.every((g) => g.granule_id.startsWith('active')));
```
Comment (Nnaga1) [inline packages/db/tests/lib/test-duplicate-detection.js:233]: why does this check what the granuleId starts with?: `g.granule_id.startsWith('active')` instead of 'state'?

File: packages/db/tests/lib/test-dupliate-detection.js (hunk 8 above)
Comment (Jkovarik) [reply packages/db/tests/lib/test-dupliate-detection.js]: Nit: consider 

```
Array.from({ length: count }, (_, i) => 
```

- Removes the map
- Doesn't allocate two arrays. 

It's super minor, but the above form is generally a better way to do a .from for iteration.

File: packages/db/src/lib/duplicate-detection.ts [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+    collectionCumulusId = pgCollection.cumulus_id;
+  }
+
+  // Query for active granules with matching producerGranuleId
+  const duplicates = await knex('granules')
+    .leftJoin('granule_groups', 'granules.cumulus_id', 'granule_groups.granule_cumulus_id')
+    .where('granules.producer_granule_id', producerGranuleId)
+    .modify((qb) => {
+      if (granuleId) {
+        qb.whereNot('granules.granule_id', granuleId);
```
Comment (Jkovarik) [inline packages/db/src/lib/duplicate-detection.ts:87]: I'm assuming from the tests the intent here is to allow providing a single exclusion granuleId if you're looking for all other records for <granuleId> versus just a group?    That seems like a good thing to have!      See comments at top of function re: docstring. 

Nit: Is there ever a reasonable situation where we'd want that to be a list?

File: packages/db/src/lib/duplicate-detection.ts (hunk 11 above)
Comment (Jkovarik) [reply packages/db/src/lib/duplicate-detection.ts]: I think we should be really careful with this function name - this isn't returning duplicate granules given a producerGranuleId, it's returning *active* duplicate granules.   Granules in an inactive state are still duplicate granule objects. 

Can we rename this method to reflect that intent?

File: packages/db/tests/search/test-GranuleSearch.js [hunk 14]
```diff
@@ -223,6 +224,10 @@ test.before(async (t) => {
   );
 });
 
+test.after.always(async (t) => {
```
Comment (Jkovarik) [inline packages/db/tests/search/test-GranuleSearch.js:226]: Good change!       We always should be doing this.

**Review by Jkovarik — CHANGES_REQUESTED** [review]
Good work Jenny - I left some comments, let me know what you think!

File: packages/db/tests/lib/test-dupliate-detection.js (hunk 8 above)
Comment (Nnaga1) [reply packages/db/tests/lib/test-dupliate-detection.js]: ok, so more performant, TIL, I'll resolve this 👍

Comment (jennyhliu) [reply packages/db/tests/lib/test-dupliate-detection.js]: [bd1509f](https://github.com/nasa/cumulus/pull/4077/commits/bd1509f4772ff2bc7b0b301b3a068069826e7484)

File: packages/db/src/lib/duplicate-detection.ts (hunk 11 above)
Comment (jennyhliu) [reply packages/db/src/lib/duplicate-detection.ts]: Added docstring and rename function and types [c6638bb](https://github.com/nasa/cumulus/pull/4077/commits/c6638bbcdd59ae192b78df8c5df75177693e9ec3)

File: packages/db/src/lib/duplicate-detection.ts (hunk 5 above)
Comment (jennyhliu) [reply packages/db/src/lib/duplicate-detection.ts:85]: This is to exclude the granule from the input if provided, and there is case that the input granule doesn't have a granuleId

File: packages/db/tests/lib/test-duplicate-detection.js (hunk 10 above)
Comment (jennyhliu) [reply packages/db/tests/lib/test-duplicate-detection.js:102]: We could check the duplicate granules found all have this prefix, but in this test case, it's redundant.

File: packages/db/tests/lib/test-duplicate-detection.js (hunk 12 above)
Comment (jennyhliu) [reply packages/db/tests/lib/test-duplicate-detection.js:233]: ~Currently we don't retrieve granule state information~ Added [c6638bb](https://github.com/nasa/cumulus/pull/4077/commits/c6638bbcdd59ae192b78df8c5df75177693e9ec3)

File: packages/db/src/lib/duplicate-detection.ts [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+  } = input;
+
+  let { collectionCumulusId } = input;
+
+  // Ensure we have collectionCumulusId
+  if (!collectionCumulusId) {
+    const pgCollection = await collectionPgModel.get(
+      knex,
+      deconstructCollectionId(collectionId)
+    );
```
Comment (jennyhliu) [inline packages/db/src/lib/duplicate-detection.ts:77]: Actually, both deconstructCollectionId and collectionPgModel.get will throw error if collectionId is not valid format, or the collection doesn't exist.

File: packages/db/src/lib/duplicate-detection.ts (hunk 4 above)
Comment (Jkovarik) [reply packages/db/src/lib/duplicate-detection.ts:6]: https://github.com/nasa/cumulus/pull/4077#discussion_r2395557220 given this, we can close this concern.

File: packages/db/src/lib/duplicate-detection.ts (hunk 9 above)
Comment (Jkovarik) [reply packages/db/src/lib/duplicate-detection.ts]: Thanks for making the change!

File: packages/db/src/lib/duplicate-detection.ts (hunk 2 above)
Comment (Jkovarik) [reply packages/db/src/lib/duplicate-detection.ts]: "  /**
   * Granules that match the input granule based on custom-defined duplication criteria.
   * Currently unused and always returns an empty array; reserved for future enhancements.
   " 
   
   Resolves this concern

File: packages/db/src/lib/duplicate-detection.ts (hunk 3 above)
Comment (Jkovarik) [reply packages/db/src/lib/duplicate-detection.ts:39]: Talked in huddle, agreed to leave this here for now as the model doesn't really make sense, and the associate lib isn't particularly better.

File: packages/db/src/lib/duplicate-detection.ts (hunk 7 above)
Comment (Jkovarik) [reply packages/db/src/lib/duplicate-detection.ts]: This was done and looks good 👍🏻

File: packages/db/src/lib/duplicate-detection.ts (hunk 11 above)
Comment (Jkovarik) [reply packages/db/src/lib/duplicate-detection.ts]: Looks great, thanks for addressing the comment!

File: packages/db/tests/lib/test-duplicate-detection.js (hunk 1 above)
Comment (Jkovarik) [reply packages/db/tests/lib/test-duplicate-detection.js:1]: I think this still needs done.

File: packages/db/tests/lib/test-duplicate-detection.js [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+
+  const id1 = await getNextGranuleGroupId(knex);
+  const id2 = await getNextGranuleGroupId(knex);
+
+  t.true(Number.isInteger(id1));
+  t.true(Number.isInteger(id2));
+  t.true(id2 > id1, 'Second call should return a higher group_id');
+});
+
+test('findActiveDuplicateGranules finds multiple duplicates in same collection', async (t) => {
```
Comment (Jkovarik) [inline packages/db/tests/lib/test-duplicate-detection.js:91]: These tests look good, a couple of minor comments:  

- Can we add tests to cover the expected behavior for empty producerGranuleId/collectionId?       We agreed that we don't need to cover validation as the downstream methods will throw, but it seems like it's a good idea to cover that expectation in a test. 
- Can we add a test to cover the return  if no records in the table and/or tests when a query happens and there's no record in the table?

File: packages/db/tests/lib/test-duplicate-detection.js [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+  });
+
+  const result = await findActiveDuplicateGranules({
+    collectionId,
+    producerGranuleId,
+    collectionCumulusId,
+  }, knex);
+
+  t.is(result.sameCollectionMatches.length, 3);
+  t.is(result.differentCollectionMatches.length, 0);
```
Comment (Jkovarik) [inline packages/db/tests/lib/test-duplicate-detection.js]: Nit: 

We're not explicitly calling out the other default behaviors here (things like customCriteriaMatches always being 0).   Given its' a feature that seems mostly ok, but there's a possible concern that when an update is made we neglect to address all existing test cases.       I'm ok leaving this for our review process ,but could also see it falling to being explicit in this PR as well.

File: packages/db/tests/lib/test-duplicate-detection.js (hunk 16 above)
Comment (jennyhliu) [reply packages/db/tests/lib/test-duplicate-detection.js:91]: Added more tests [1fe9a51](https://github.com/nasa/cumulus/pull/4077/commits/1fe9a51c1ae5680b53f1487b48c1219c67a2c5db)

**Review by Jkovarik — APPROVED** [review]
@jennyhliu  the changes look good, great work!      Thanks for addressing/discussing the comments!
